### DISCUSSION TITLE
Socket.to is not a function

### DISCUSSION THREAD

**Discussion by pepper28** [socketio/socket.io#4519] [Q&A] [answered]
I have employed the response to the client socket method as socket.to(sessionID).emit(txId);  and I received the following error message.  TypeError: socket.to is not a function.  Everything I have read has said this is the appropriate way to send the response from the server to a specific client.  Please HELP!
          ^

**Comment by darrachequesne** [comment]
Hi! That's weird, the `to()` method should indeed exist. Could you please provide a sample project reproducing the issue?

**Comment by pepper28** [comment, marked as answer]
I've tried the following "socket.to(sessionID).emit(txId);", "socket.emit(txId);", and "server.to(sessionID).emit(txId);".  My client is a java machine and although I'm sending userInfo and points successfully using, "socketWriter = new BufferedWriter(new OutputStreamWriter(socket.getOutputStream())); " to Nodejs socket.io.  I can only receive the txId if the method  I'm using is socket.to(sessionID).emit(txId), but this crashes Nodejs.  If I use socket.emit(txId) Java client running  " socketReader = new BufferedReader(new InputStreamReader(socket.getInputStream()));" never gets call or txId data.

My environment for both the Java client and the NodeJS Server is Ubuntu Linux version 18.04.


Here is my Node JS code 

```
const server = require('net').createServer();
const RPCClient = require('');
const { SocketAddress } = require('net');
const { Socket } = require('socket.io');
const avcCRC = new RPCClient('');
const socket = require('socket.io')(server);

server.on('connection', socket =>{
  console.log('Client connected');
  const sessionID = socket.id;   
  var remoteAddress = socket.remoteAddress + ":" + socket.remotePort
 socket.on('data',data => { 
   //socket.write(data);
   const commaIndex = data.indexOf(",");
   userInfo = data.substring(0, commaIndex);
   console.log(userInfo);
   points = data.substring(commaIndex +1);
   console.log(points);
  
   avcCRC.call('passphrase', 'tester28', 15).then(() => {
   avcCRC.call('senduserInfo', userInfo, points).then(txId => {
    
  socket.to(sessionID).emit(txId);
  //socket.emit(txId);
  
   console.log('The result is:  ' + txId );     
});
});
});
            
socket.setEncoding('utf8');
socket.once("close", function(){
console.log("connection from %s closed", remoteAddress);
});

socket.on("error", function(err){
  console.log("connection %s error: %s", remoteAddress, err.message);
});

});

   server.listen(9000, () => console.log('server is listening on port 9000'));
```

Edit by @darrachequesne: format

**Reply by darrachequesne** [reply]
Hi! I think you need to pass a HTTP server (`require('http').createServer()`), and not a TCP server (`require('net').createServer()`):

```js
const { createServer } = require("http");
const { Server } = require("socket.io");

const httpServer = createServer();
const io = new Server(httpServer, { /* options */ });

io.on("connection", (socket) => {
  // ...
});

httpServer.listen(3000);
```

Reference: https://socket.io/docs/v4/server-initialization/#with-an-http-server

**Reply by pepper28** [reply]
Damien, I want to thank you.  Although I thought using the base TCP networking method would have been sufficient your resolution solved my issue.  Thank you again.

**Comment by pepper28** [comment]
I have written a very minimal NodeJS stub to see if any of the code was interfering with "sockect.to' operating.  Even in this minimal configuration, I am receiving the following error.

Process exited with code 1
Uncaught TypeError TypeError: socket.to is not a function
    at <anonymous> (/home/boulder/BCNode/interop5.js:17:11)
    at emit (events:513:28)
    at addChunk (internal/streams/readable:315:12)
    at readableAddChunk (internal/streams/readable:285:11)
    at Readable.push (internal/streams/readable:228:10)
    at onStreamRead (internal/stream_base_commons:190:23)
    at callbackTrampoline (internal/async_hooks:130:17)
    --- TCPWRAP ---
    at init (internal/inspector_async_hook:25:19)
    at emitInitNative (internal/async_hooks:202:43)
No debugger available, can not send 'variables'

Here is the minimal code I used.

const server = require('net').createServer();
const socket = require('socket.io')(server);

server.on('connection', socket =>{
  console.log('Client connected');
  const sessionID = socket.id;   
 socket.on('data',data => { 
   socket.write(data);
   const commaIndex = data.indexOf(",");
   userInfo = data.substring(0, commaIndex);
   console.log(userInfo);
   points = data.substring(commaIndex +1);
   console.log(points);
  
   socket.to(sessionID).emit("Trying to send info to the client"); 
  
});

socket.setEncoding('utf8');
});

server.listen(9000, () => console.log('server is listening on port 9000'));
